tests/integration_tests.rs

```diff
@@ -0,0 +1,643 @@
+    );
+    Ok(())
+}
+
+#[tokio::test]
+async fn untrusted_server() -> Result<()> {
+    let _guard = subscribe();
+
+    // Run the server with an unknown CA.
+    let mut server_crypto = ServerConfig::new().unwrap();
+    server_crypto
+        .set_cert(UNTRUSTED_SERVER_CERT.chain(), UNTRUSTED_SERVER_CERT.key())
+        .unwrap();
+    for cert in CLIENT_CERT.chain() {
+        server_crypto.add_trusted_cert(cert).unwrap();
+    }
+    let server = Server::run(server_config(server_crypto)?)?;
+
+    // Create the client with no server certs configured.
+    let err = Client::connect(client_config(client_crypto()), server.addr)
+        .await
+        .expect_err("expected connection failure");
+    match err {
+        ConnectionError::TransportError(e) => {
+            assert_eq!(TransportErrorCode::crypto(UNKNOWN_CA), e.code);
+            Ok(())
+        }
+        _ => Err(anyhow!("unexpected error: {err}")),
+    }
+}
+
+#[tokio::test]
+async fn untrusted_client() -> Result<()> {
+    let _guard = subscribe();
+
+    // Run the server with client auth enabled.
+    let mut server_crypto = server_crypto();
+    server_crypto.enable_client_auth(true);
+    let server = Server::run(server_config(server_crypto)?)?;
+
+    // Create the client with an unknown CA.
+    let mut client_crypto = ClientConfig::new().unwrap();
+    client_crypto
+        .set_cert(UNTRUSTED_CLIENT_CERT.chain(), UNTRUSTED_CLIENT_CERT.key())
+        .unwrap();
+    for cert in SERVER_CERT.chain() {
+        client_crypto.add_trusted_cert(cert).unwrap();
+    }
+    let client = Client::connect(client_config(client_crypto), server.addr).await?;
+    let (send, _) = client.conn.open_bi().await?;
+    let request = &b"hello world"[..];
+    let err = Client::send_request(send, request)
+        .await
+        .expect_err("expected connection failure");
+    match err {
+        WriteError::ConnectionLost(ConnectionError::ConnectionClosed(e)) => {
+            assert_eq!(TransportErrorCode::crypto(UNKNOWN_CA), e.error_code);
+            Ok(())
+        }
+        _ => Err(anyhow!("unexpected error: {err}")),
+    }
+}
+
+#[tokio::test]
+async fn zero_rtt_success() -> Result<()> {
+    let _guard = subscribe();
+
+    // Run the server.
+    let server = Server::run(server_config(server_crypto())?)?;
+
+    let cc1 = client_crypto();
+    let session_cache = cc1.get_session_cache();
+    let client1 = Client::connect_no_zero_rtt(client_config(cc1), server.addr).await?;
+    client1.ping_pong().await?;
+
+    info!("initial connection complete");
+
+    // Create the second client with the shared session cache.
+    let mut cc2 = client_crypto();
+    cc2.set_session_cache(session_cache);
+    let (client2, zero_rtt) = Client::connect_zero_rtt(client_config(cc2), server.addr).await?;
+    client2.ping_pong().await?;
+    assert!(zero_rtt.await, "expected 0-RTT");
+    Ok(())
+}
+
+#[tokio::test]
+async fn zero_rtt_rejected() -> Result<()> {
```

Comment:
@djc this test passes, but takes ~10s, which makes me think that it's hitting some timeout. I noticed that [quinn's zero_rtt test](https://github.com/quinn-rs/quinn/blob/7424372f49f285d4164facc42e069b4226310fa2/quinn/src/tests.rs#L253) also takes ~10s, which makes me think that it's hitting the same timeout. Any thoughts?

Reply:
Sounds iffy -- not sure what's going on there. If you want to dig into that, would be great!

Reply:
@djc I was able to avoid the 10s delay (due to the connection timeout) by adding the sleep below.

Reply:
Did you figure out why that was necessary? (If it should be necessary?)